### PR TITLE
Added new functionality of creating and checking tokens

### PR DESCRIPTION
**Pull request by CapUlDis**

### CODE REVIEW

File: auth.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        raise TypeError(f'InvalidJWKValue: data in {path_token_key} is not JWK object')   
+    
+class Token_gen:
+
+    def __init__(self, path_token_key):
+        self.load_key = load_token_key(path_token_key)
+
+    def create_new_token(self, username, token_exp):
+        token = jwt.JWT(header={"alg": "HS256"},
+                        claims={"user": username, "exp": calendar.timegm((datetime.utcnow() + timedelta(minutes = token_exp)).timetuple())})
```
Comment (nkartashov) [inline auth.py:27]: Why not `int((datetime.utcnow() + timedelta(minutes = token_exp)).timestamp())` ?

File: main.py [hunk 2]
```diff
@@ -31,16 +32,71 @@ def login():
     if not current_app.credit.check_user_with_password_exists(data['name'], data['password']):
         return 'Invalid name or password.', 403
         
-    return 'Correct name and password.', 200
+    return jsonify({'token': current_app.token_gen.create_new_token(data['name'], current_app.token_exp)}), 200
+
+def sign_in():
```
Comment (nkartashov) [inline main.py]: `sign_in()` is very similar in meaning to login, consider renaming it to reflect what you return

File: main.py [hunk 3]
```diff
@@ -31,16 +32,71 @@ def login():
...
+    
+    try:
+        token = request.headers['Authorization']
+    except KeyError as inf:
+        logger.info(f'In post request Authorization header missing: {inf}')
+        return 'Please log in again', 401
```
Comment (nkartashov) [inline main.py]: Would logging in help with a missing header?

File: main.py [hunk 4]
```diff
@@ -31,16 +32,71 @@ def login():
...
     
+    token = token.replace('Bearer ', '')
+
+    if token == '':
+        logger.info('In Authorization header token missing')
+        return 'Please log in again', 401
```
Comment (nkartashov) [inline main.py]: Would logging in help with a malformed header?

File: main.py [hunk 5]
```diff
@@ -31,16 +32,71 @@ def login():
...
+
+    if not current_app.token_gen.check_token(token):
+        logger.info('Got invalid token')
+        return 'Please log in again', 401
+
+    return jsonify({
+        'userName': 'denchik',
+        'id': 1,
+        'numOfLists': 1,
```
Comment (nkartashov) [inline main.py]: Is there a point to send the number of lists back if you send them back in the same message?

File: main.py [hunk 6]
```diff
@@ -31,16 +32,71 @@ def login():
...
+        'toDoLists': {
+            'listName': 'First to-do list',
+            'listDate': '19 Dec 2019',
+            'listColor': 'green',
+            'listItems': [
+                {
+                    'itemNum': 1,
```
Comment (nkartashov) [inline main.py:70]: Why would we send the `itermNum` here?

File: main.py [hunk 7]
```diff
@@ -31,16 +32,71 @@ def login():
...
+        logger.info('Got invalid token')
+        return 'Please log in again', 401
+
+    return jsonify({
+        'userName': 'denchik',
+        'id': 1,
+        'numOfLists': 1,
+        'toDoLists': {
+            'listName': 'First to-do list',
+            'listDate': '19 Dec 2019',
```
Comment (nkartashov) [inline main.py]: What does this date represent?

File: main.py [hunk 8]
```diff
@@ -31,16 +32,71 @@ def login():
...
+            'listColor': 'green',
+            'listItems': [
+                {
+                    'itemNum': 1,
+                    'itemName': 'Buy some milk',
+                    'itemTerm': 'today',
```
Comment (nkartashov) [inline main.py]: What does `itemTerm` mean?

File: tests/test_auth.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    ('n_case', 'expected_str'),
+    (
+        (0, 'JWTExpired: token is expired'),
+        (1, 'InvalidJWSSignature: token has invalid signature'),
+        (2, 'ValueError: received string is not JSON Web Token'),
+        (3, 'InvalidJWSObject: invalid token format')
+    ),
+    ids = ['Expired token', 'Token has invalid signature', 'Received string is not JSON Web Token', 'Invalid token format']
+)    
+def test_class_Tokem_check_token_all_error_cases(n_case, expected_str):
```
Comment (nkartashov) [inline tests/test_auth.py]: Tokem -> Token

File: auth.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            token_key = jwk.JWK.from_json(f.read())
+            return token_key
+    except FileNotFoundError as err:
+        logger.error(f'FileNotFoundError: no credentials.txt in {path_token_key}: {err}')
+        raise FileNotFoundError(f'FileNotFoundError: no credentials.txt in {path_token_key}: {err}')
+    except jwk.InvalidJWKValue as err:
+        logger.error(f'InvalidJWKValue: data in {path_token_key} is not JWK object')     
+        raise TypeError(f'InvalidJWKValue: data in {path_token_key} is not JWK object')   
+    
+class Token_gen:
```
Comment (nkartashov) [inline auth.py]: You probably want to rename this class to match python naming conventions (`TokenGen` or better `TokenGenerator`)

File: tests/test_auth.py (hunk 9 above)
Comment (nkartashov) [inline tests/test_auth.py]: Why not split this test into several, there is very little reuse happening. Alternatively, you can move the token initialisation into the parameters of your test.

File: tests/test_auth.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    assert expected_str in mock_error.call_args[0][0]
+
+def test_load_key_normal_case():
+    test_token_key = load_token_key(os.path.dirname(os.path.realpath(__file__)) + '/test_token_key.txt')
+    assert test_token_key.export() == '{"k":"foo","kty":"oct"}'
+
+def test_class_Token_create_and_check_token_normal_case():
+    test_token_object  = Token_gen(os.path.dirname(os.path.realpath(__file__)) + '/test_token_key.txt')
+    str_token = test_token_object.create_new_token('testname', 30)
+    assert str_token.count('.') == 2
```
Comment (nkartashov) [inline tests/test_auth.py]: Why do we want to have this check?

File: tests/test_post_sign_in.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+)
+def test_sign_in_error_cases(client, app, headers, logger_message, http_code):
+    with mock.patch.object(logger, 'info') as mock_info:
+        response_error = client.post('/my-todos', headers = headers)
+        assert logger_message in mock_info.call_args[0][0]
+        assert http_code == response_error.status_code
+
+def test_sign_in_valid_token(client, app):
+    response_login = client.post('/login', data = json.dumps({"name": "denchik", "password": "foobar"}), headers = {'content-type': 'application/json'})
+    recieved_token = json.loads(response_login.data)
```
Comment (nkartashov) [inline tests/test_post_sign_in.py]: recieved -> received

File: tests/test_post_sign_in.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    response_sign_in = client.post('/my-todos', headers = {'Authorization': recieved_token['token']})    
+    assert 200 == response_sign_in.status_code
+    assert 'toDoLists' in str(response_sign_in.data)
```
Comment (nkartashov) [inline tests/test_post_sign_in.py]: You can parse the response and compare directly.

**Review by nkartashov — CHANGES_REQUESTED** [review]
Very good progress! I don't have any major code improvements to suggest, great work!

File: auth.py (hunk 1 above)
Comment (CapUlDis) [reply auth.py:27]: That operation returns timestamp, which is 3 hours before given.

File: main.py (hunk 2 above)
Comment (CapUlDis) [reply main.py]: done!

File: main.py (hunk 7 above)
Comment (CapUlDis) [reply main.py]: This date represents when list was created.

File: main.py (hunk 8 above)
Comment (CapUlDis) [reply main.py]: `itemTerm` means deadline of task.

File: auth.py (hunk 10 above)
Comment (CapUlDis) [reply auth.py]: Done!

File: tests/test_post_sign_in.py (hunk 12 above)
Comment (CapUlDis) [reply tests/test_post_sign_in.py]: Done.

File: tests/test_auth.py (hunk 9 above)
Comment (CapUlDis) [reply tests/test_auth.py]: Done.

File: main.py (hunk 8 above)
Comment (CapUlDis) [reply main.py]: Changed to `itemDeadline`.

File: main.py (hunk 6 above)
Comment (CapUlDis) [reply main.py:70]: `itemNum` represents the number of this list's item.

File: main.py (hunk 5 above)
Comment (CapUlDis) [reply main.py]: But what if program returns not that number of created lists?

File: tests/test_auth.py (hunk 11 above)
Comment (CapUlDis) [reply tests/test_auth.py]: Done!

File: main.py (hunk 5 above)
Comment (CapUlDis) [reply main.py]: Done!

File: main.py (hunk 4 above)
Comment (CapUlDis) [reply main.py]: Done!

File: main.py (hunk 3 above)
Comment (CapUlDis) [reply main.py]: Done!

File: tests/test_auth.py (hunk 9 above)
Comment (CapUlDis) [reply tests/test_auth.py]: Done!

File: tests/test_post_sign_in.py (hunk 13 above)
Comment (CapUlDis) [reply tests/test_post_sign_in.py]: Done!

**Review by nkartashov — APPROVED** [review]

**Pull request merged**